File: server/src/quotas/gemini.ts
// server/src/quotas/gemini.ts — Gemini CLI（oauth-personal）订阅额度（移植 CodexBar GeminiStatusProbe）：
// 凭据 = ~/.gemini/oauth_creds.json（access_token/refresh_token/expiry_date），settings.json 的
// selectedType 不能是 api-key；过期时用 refresh_token + OAuth client（env 或从本机 gemini-cli
// 安装文件里抽取）刷新并回写；loadCodeAssist → retrieveUserQuota，buckets 按模型取最低剩余比例。
import { existsSync, readFileSync, readdirSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import {
  finalizeAccount, missingAccount, quotaFetchJson, readHttpStatus, errText,
  quotaNum, quotaStr, toEpochMs,
  type FetchLike, type QuotaAccount, type QuotaWindow,
} from './types.js'

const LABEL = 'Gemini'

export type GeminiCreds = {
  accessToken: string | null
  refreshToken: string | null
  expiryMs: number | null
  /** 已解析/抽取到的 OAuth client；用于静默刷新 */
  clientId: string | null
  clientSecret: string | null
  credsPath: string
}

const TIER_NAMES: Record<string, string> = {
  'free-tier': 'Free',
  'legacy-tier': 'Legacy',
  'standard-tier': 'Standard',
}

/** gemini-cli 安装里 oauth2.js 的候选路径（有 GEMINI_OAUTH2_JS_PATH/GEMINI_CLI_HOME 则优先） */
export function oauth2JsCandidates(home: string, env: NodeJS.ProcessEnv): string[] {
  const direct = env['GEMINI_OAUTH2_JS_PATH']?.trim()
  const cliHome = env['GEMINI_CLI_HOME']?.trim()
  const fixed = [
    '/opt/homebrew/lib/node_modules/@google/gemini-cli',
    '/usr/local/lib/node_modules/@google/gemini-cli',
    join(home, '.bun/install/global/node_modules/@google/gemini-cli'),
  ]
  const roots: string[] = []
  if (cliHome) roots.push(cliHome)
  try {
    const nvmDir = join(home, '.nvm', 'versions', 'node')
    for (const version of readdirSync(nvmDir)) {
      roots.push(join(nvmDir, version, 'lib', 'node_modules', '@google', 'gemini-cli'))
    }
  } catch {
    // 无 nvm
  }
  roots.push(...fixed)
  const candidates: string[] = []
  if (direct) candidates.push(direct)
  for (const root of roots) {
    candidates.push(join(root, 'dist', 'src', 'code_assist', 'oauth2.js'))
    candidates.push(join(root, 'src', 'code_assist', 'oauth2.js'))
  }
  return candidates
}

export function extractOAuthClient(home: string, env: NodeJS.ProcessEnv): { clientId: string; clientSecret: string } | null {
  const clientId = env['GEMINI_OAUTH_CLIENT_ID']?.trim()
  const clientSecret = env['GEMINI_OAUTH_CLIENT_SECRET']?.trim()
  if (clientId && clientSecret) return { clientId, clientSecret }
  for (const path of oauth2JsCandidates(home, env)) {
    try {
      if (!existsSync(path)) continue
      const content = readFileSync(path, 'utf8')
      const id = content.match(/(?:client_id|CLIENT_ID)['"`]?\s*[:=]\s*['"]([^'"]+)['"]/)?.[1]
      const secret = content.match(/(?:client_secret|CLIENT_SECRET)['"`]?\s*[:=]\s*['"]([^'"]+)['"]/)?.[1]
      if (id && secret) return { clientId: id, clientSecret: secret }
    } catch {
      // 换下一个候选路径
    }
  }
  return null
}

export function readGeminiCreds(home: string, env: NodeJS.ProcessEnv): GeminiCreds | null {
  // settings.json 明确选了 API key 模式 → 没有订阅额度可看
  try {
    const settings = JSON.parse(readFileSync(join(home, '.gemini', 'settings.json'), 'utf8')) as Record<string, any>
    if (settings?.['security']?.['auth']?.['selectedType'] === 'api-key') return null
  } catch {
    // 无 settings.json（老版本默认 oauth-personal）→ 继续
  }
  const credsPath = join(home, '.gemini', 'oauth_creds.json')
  let parsed: Record<string, any>
  try {
    parsed = JSON.parse(readFileSync(credsPath, 'utf8')) as Record<string, any>
  } catch {
    return null
  }
  if (typeof parsed !== 'object' || parsed === null) return null
  const client = extractOAuthClient(home, env)
  return {
    accessToken: quotaStr(parsed['access_token']),
    refreshToken: quotaStr(parsed['refresh_token']),
    expiryMs: quotaNum(parsed['expiry_date']),
    clientId: client?.clientId ?? null,
    clientSecret: client?.clientSecret ?? null,
    credsPath,
  }
}

/** access_token 过期（< now+60s）时尝试静默刷新并回写 oauth_creds.json；失败返回原值由 API 报 401 */
export async function ensureFreshAccessToken(
  creds: GeminiCreds, doFetch: FetchLike, now: number,
): Promise<string | null> {
  const fresh = creds.accessToken !== null && (creds.expiryMs === null || creds.expiryMs > now + 60_000)
  if (fresh) return creds.accessToken
  if (!creds.refreshToken || !creds.clientId || !creds.clientSecret) return creds.accessToken
  try {
    const res = await doFetch('https://oauth2.googleapis.com/token', {
      method: 'POST',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({
        client_id: creds.clientId,
        client_secret: creds.clientSecret,
        refresh_token: creds.refreshToken,
        grant_type: 'refresh_token',
      }).toString(),
      signal: AbortSignal.timeout(15_000),
    })
    if (!res.ok) return creds.accessToken
    const json = (await res.json()) as Record<string, unknown>
    const newToken = quotaStr(json['access_token'])
    if (!newToken) return creds.accessToken
    const expiresIn = quotaNum(json['expires_in'])
    try {
      const updated = JSON.parse(readFileSync(creds.credsPath, 'utf8')) as Record<string, any>
      updated['access_token'] = newToken
      if (expiresIn !== null) updated['expiry_date'] = now + expiresIn * 1000
      if (quotaStr(json['id_token'])) updated['id_token'] = json['id_token']
      writeFileSync(creds.credsPath, JSON.stringify(updated, null, 2))
    } catch {
      // 回写失败不影响本轮读取
    }
    return newToken
  } catch {
    return creds.accessToken
  }
}

type CodeAssistStatus = { tier: string | null; projectId: string | null }

async function loadCodeAssistStatus(token: string, doFetch: FetchLike): Promise<CodeAssistStatus> {
  const { body } = await quotaFetchJson(
    doFetch, 'https://cloudcode-pa.googleapis.com/v1internal:loadCodeAssist', {
      method: 'POST',
      headers: { authorization: `Bearer ${token}`, 'content-type': 'application/json' },
      body: JSON.stringify({ metadata: { ideType: 'GEMINI_CLI', pluginType: 'GEMINI' } }),
    },
  )
  const root = (body ?? {}) as Record<string, any>
  const projectRaw = root['cloudaicompanionProject']
  return {
    tier: quotaStr(root['currentTier']),
    projectId: quotaStr(projectRaw) ?? quotaStr((projectRaw as Record<string, any> | null)?.['id']),
  }
}

/** 兜底项目发现：gen-lang-client* 前缀或带 generative-language 标签的项目 */
async function discoverProjectId(token: string, doFetch: FetchLike): Promise<string | null> {
  try {
    const { body } = await quotaFetchJson(
      doFetch, 'https://cloudresourcemanager.googleapis.com/v1/projects', {
        headers: { authorization: `Bearer ${token}`, accept: 'application/json' },
      },
    )
    const projects = ((body ?? {}) as Record<string, any>)['projects']
    if (!Array.isArray(projects)) return null
    for (const project of projects) {
      const id = quotaStr((project as Record<string, any>)?.['projectId'])
      if (!id) continue
      if (id.startsWith('gen-lang-client')) return id
      const labels = (project as Record<string, any>)?.['labels']
      if (labels && typeof labels === 'object' && 'generative-language' in labels) return id
    }
  } catch {
    // 项目发现是尽力而为
  }
  return null
}

async function retrieveUserQuota(
  token: string, projectId: string | null, doFetch: FetchLike,
): Promise<Array<Record<string, any>>> {
  const body: Record<string, unknown> = projectId ? { project: projectId } : {}
  const res = await quotaFetchJson(
    doFetch, 'https://cloudcode-pa.googleapis.com/v1internal:retrieveUserQuota', {
      method: 'POST',
      headers: { authorization: `Bearer ${token}`, 'content-type': 'application/json' },
      body: JSON.stringify(body),
    },
  )
  const buckets = ((res.body ?? {}) as Record<string, any>)['buckets']
  return Array.isArray(buckets) ? buckets : []
}

export async function fetchGeminiAccount(
  creds: GeminiCreds, doFetch: FetchLike, now: number,
): Promise<QuotaAccount> {
  try {
    const token = await ensureFreshAccessToken(creds, doFetch, now)
    if (!token) throw new Error('oauth_creds.json 缺少 access_token 且无法刷新')

    let status: CodeAssistStatus = { tier: null, projectId: null }
    try {
      status = await loadCodeAssistStatus(token, doFetch)
    } catch {
      // loadCodeAssist 失败不阻塞配额查询
    }
    const projectId = status.projectId ?? await discoverProjectId(token, doFetch)
    const buckets = await retrieveUserQuota(token, projectId, doFetch)
    if (buckets.length === 0) throw new Error('配额响应缺少 buckets 数据')

    // 每个模型取最低 remaining_fraction 的桶（通常 input 侧先到限）
    const perModel = new Map<string, { fraction: number; resetAt: number | null }>()
    for (const raw of buckets) {
      if (raw === null || typeof raw !== 'object') continue
      const bucket = raw as Record<string, unknown>
      const modelId = quotaStr(bucket['model_id'])
      const fraction = quotaNum(bucket['remaining_fraction'])
      if (!modelId || fraction === null) continue
      const existing = perModel.get(modelId)
      if (!existing || fraction < existing.fraction) {
        perModel.set(modelId, { fraction, resetAt: toEpochMs(bucket['reset_time']) })
      }
    }
    const windows: QuotaWindow[] = [...perModel.entries()]
      .map(([modelId, info]) => ({
        key: modelId,
        label: modelId,
        usedPercent: Math.min(100, Math.max(0, (1 - info.fraction) * 100)),
        percentage: Math.min(100, Math.max(0, (1 - info.fraction) * 100)),
        nextResetAt: info.resetAt,
      }))
      .sort((a, b) => (b.percentage ?? 0) - (a.percentage ?? 0))
      .slice(0, 6)

    return finalizeAccount({
      kind: 'gemini', label: LABEL, provider: null,
      planName: status.tier ? TIER_NAMES[status.tier] ?? status.tier : null,
      windows, anySuccess: true, now,
    })
  } catch (err) {
    return finalizeAccount({
      kind: 'gemini', label: LABEL, provider: null,
      error: errText(err), httpStatus: readHttpStatus(err), now,
    })
  }
}

export async function geminiAccount(home: string, env: NodeJS.ProcessEnv, doFetch: FetchLike, now: number): Promise<QuotaAccount> {
  const creds = readGeminiCreds(home, env)
  if (!creds) return missingAccount('gemini', LABEL, now)
  return fetchGeminiAccount(creds, doFetch, now)
}

File: Sources/WattsonCore/Quotas/Gemini.swift
// Quotas/Gemini.swift — Gemini CLI（oauth-personal）订阅额度：凭据 = ~/.gemini/oauth_creds.json
// （access_token/refresh_token/expiry_date），settings.json 的 selectedType 不能是 api-key；
// 过期时用 refresh_token + OAuth client（env 或从本机 gemini-cli 安装文件里抽取）刷新并回写；
// loadCodeAssist → retrieveUserQuota，buckets 按模型取最低剩余比例。
import Foundation

public struct GeminiCreds: Sendable, Equatable {
    public var accessToken: String?
    public var refreshToken: String?
    public var expiryMs: Double?
    /// 已解析/抽取到的 OAuth client；用于静默刷新
    public var clientId: String?
    public var clientSecret: String?
    public var credsPath: String
    public init(accessToken: String?, refreshToken: String?, expiryMs: Double?,
                clientId: String?, clientSecret: String?, credsPath: String) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiryMs = expiryMs
        self.clientId = clientId
        self.clientSecret = clientSecret
        self.credsPath = credsPath
    }
}

private let geminiLabel = "Gemini"

private let geminiTierNames: [String: String] = [
    "free-tier": "Free",
    "legacy-tier": "Legacy",
    "standard-tier": "Standard",
]

/// 纯文本错误（无 HTTP 状态码）
private struct GeminiError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

/// 正则第 1 个捕获组（TS content.match(re)?.[1] 口径）
private func firstRegexCapture(_ pattern: String, _ content: String) -> String? {
    guard let re = try? NSRegularExpression(pattern: pattern) else { return nil }
    let range = NSRange(content.startIndex..., in: content)
    guard let m = re.firstMatch(in: content, options: [], range: range),
          m.numberOfRanges > 1, let captured = Range(m.range(at: 1), in: content) else { return nil }
    return String(content[captured])
}

/// gemini-cli 安装里 oauth2.js 的候选路径（有 GEMINI_OAUTH2_JS_PATH/GEMINI_CLI_HOME 则优先）
public func oauth2JsCandidates(_ home: String, _ env: [String: String]) -> [String] {
    let direct = env["GEMINI_OAUTH2_JS_PATH"]?.trimmingCharacters(in: .whitespaces)
    let cliHome = env["GEMINI_CLI_HOME"]?.trimmingCharacters(in: .whitespaces)
    let fixed = [
        "/opt/homebrew/lib/node_modules/@google/gemini-cli",
        "/usr/local/lib/node_modules/@google/gemini-cli",
        (home as NSString).appendingPathComponent(".bun/install/global/node_modules/@google/gemini-cli"),
    ]
    var roots: [String] = []
    if let cliHome { roots.append(cliHome) }
    let nvmDir = (home as NSString).appendingPathComponent(".nvm/versions/node")
    for version in listDirectory(nvmDir) {
        roots.append((nvmDir as NSString).appendingPathComponent("\(version)/lib/node_modules/@google/gemini-cli"))
    }
    roots.append(contentsOf: fixed)
    var candidates: [String] = []
    if let direct { candidates.append(direct) }
    for root in roots {
        candidates.append((root as NSString).appendingPathComponent("dist/src/code_assist/oauth2.js"))
        candidates.append((root as NSString).appendingPathComponent("src/code_assist/oauth2.js"))
    }
    return candidates
}

public func extractOAuthClient(_ home: String, _ env: [String: String]) -> (clientId: String, clientSecret: String)? {
    let clientId = env["GEMINI_OAUTH_CLIENT_ID"]?.trimmingCharacters(in: .whitespaces)
    let clientSecret = env["GEMINI_OAUTH_CLIENT_SECRET"]?.trimmingCharacters(in: .whitespaces)
    if let clientId, let clientSecret { return (clientId, clientSecret) }
    for path in oauth2JsCandidates(home, env) {
        guard fileExists(path), let content = readTextFile(path) else { continue }
        let id = firstRegexCapture("(?:client_id|CLIENT_ID)['\"`]?\\s*[:=]\\s*['\"]([^'\"]+)['\"]", content)
        let secret = firstRegexCapture("(?:client_secret|CLIENT_SECRET)['\"`]?\\s*[:=]\\s*['\"]([^'\"]+)['\"]", content)
        if let id, let secret { return (id, secret) }
    }
    return nil
}

/// 凭据：~/.gemini/oauth_creds.json；settings.json 明确选了 API key 模式 → 没有订阅额度可看
public func readGeminiCreds(_ home: String, _ env: [String: String]) -> GeminiCreds? {
    // 无 settings.json（老版本默认 oauth-personal）→ 继续
    let geminiHome = (home as NSString).appendingPathComponent(".gemini")
    if let settings = parseJSONFile((geminiHome as NSString).appendingPathComponent("settings.json")),
       settings["security"]?["auth"]?["selectedType"] == .str("api-key") {
        return nil
    }
    let credsPath = (geminiHome as NSString).appendingPathComponent("oauth_creds.json")
    guard let parsed = parseJSONFile(credsPath) else { return nil }
    guard !parsed.isNull, parsed.obj != nil || parsed.arr != nil else { return nil }
    let client = extractOAuthClient(home, env)
    return GeminiCreds(
        accessToken: quotaStr(parsed["access_token"]),
        refreshToken: quotaStr(parsed["refresh_token"]),
        expiryMs: quotaNum(parsed["expiry_date"]),
        clientId: client?.clientId,
        clientSecret: client?.clientSecret,
        credsPath: credsPath)
}

/// application/x-www-form-urlencoded 值编码（对应 TS new URLSearchParams(...).toString()）
private func formEncode(_ value: String) -> String {
    // query 字符集剔除分隔符 & = ? +：值内这些字符必须转义（+ 在表单里是空格）
    let allowed = CharacterSet.urlQueryAllowed.subtracting(CharacterSet(charactersIn: "&=?+"))
    return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
}

/// access_token 过期（< now+60s）时尝试静默刷新并回写 oauth_creds.json；失败返回原值由 API 报 401
public func ensureFreshAccessToken(_ creds: GeminiCreds, _ doFetch: FetchLike, _ now: Double) async -> String? {
    let fresh = creds.accessToken != nil && (creds.expiryMs.map { $0 > now + 60_000 } ?? true)
    if fresh, let token = creds.accessToken { return token }
    guard let refreshToken = creds.refreshToken, let clientId = creds.clientId,
          let clientSecret = creds.clientSecret else { return creds.accessToken }
    do {
        let form = [
            "client_id=\(formEncode(clientId))",
            "client_secret=\(formEncode(clientSecret))",
            "refresh_token=\(formEncode(refreshToken))",
            "grant_type=refresh_token",
        ].joined(separator: "&")
        let res = try await doFetch(FetchRequest(
            url: "https://oauth2.googleapis.com/token", method: "POST",
            headers: ["content-type": "application/x-www-form-urlencoded"], body: form))
        guard res.ok else { return creds.accessToken }
        let json = res.body
        guard let newToken = quotaStr(json["access_token"]) else { return creds.accessToken }
        let expiresIn = quotaNum(json["expires_in"])
        // 回写失败不影响本轮读取
        if var updated = parseJSONFile(creds.credsPath)?.obj {
            updated["access_token"] = .str(newToken)
            if let expiresIn { updated["expiry_date"] = .num(now + expiresIn * 1000) }
            if let rawIdToken = json["id_token"], quotaStr(rawIdToken) != nil { updated["id_token"] = rawIdToken }
            if let data = try? JSON.obj(updated).encoded(pretty: true),
               let out = String(data: data, encoding: .utf8) {
                _ = writeTextFile(creds.credsPath, out)
            }
        }
        return newToken
    } catch {
        return creds.accessToken
    }
}

private struct CodeAssistStatus {
    var tier: String?
    var projectId: String?
}

private func loadCodeAssistStatus(_ token: String, _ doFetch: FetchLike) async throws -> CodeAssistStatus {
    let result = try await quotaFetchJson(
        doFetch, "https://cloudcode-pa.googleapis.com/v1internal:loadCodeAssist", method: "POST",
        headers: ["authorization": "Bearer \(token)", "content-type": "application/json"],
        body: "{\"metadata\":{\"ideType\":\"GEMINI_CLI\",\"pluginType\":\"GEMINI\"}}")
    let root = result.body
    let projectRaw = root["cloudaicompanionProject"]
    return CodeAssistStatus(
        tier: quotaStr(root["currentTier"]),
        projectId: quotaStr(projectRaw) ?? quotaStr(projectRaw?["id"]))
}

/// 兜底项目发现：gen-lang-client* 前缀或带 generative-language 标签的项目
private func discoverProjectId(_ token: String, _ doFetch: FetchLike) async -> String? {
    do {
        let result = try await quotaFetchJson(
            doFetch, "https://cloudresourcemanager.googleapis.com/v1/projects",
            headers: ["authorization": "Bearer \(token)", "accept": "application/json"])
        guard let projects = result.body["projects"]?.arr else { return nil }
        for project in projects {
            guard let id = quotaStr(project["projectId"]) else { continue }
            if id.hasPrefix("gen-lang-client") { return id }
            if project["labels"]?.obj?["generative-language"] != nil { return id }
        }
    } catch {
        // 项目发现是尽力而为
    }
    return nil
}

private func retrieveUserQuota(_ token: String, _ projectId: String?, _ doFetch: FetchLike) async throws -> [JSON] {
    let payload = JSON.obj(projectId.map { ["project": JSON.str($0)] } ?? [:])
    let body = (try? payload.encoded()).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
    let res = try await quotaFetchJson(
        doFetch, "https://cloudcode-pa.googleapis.com/v1internal:retrieveUserQuota", method: "POST",
        headers: ["authorization": "Bearer \(token)", "content-type": "application/json"],
        body: body)
    return res.body["buckets"]?.arr ?? []
}

public func fetchGeminiAccount(_ creds: GeminiCreds, _ doFetch: FetchLike, _ now: Double) async -> QuotaAccount {
    do {
        let token = await ensureFreshAccessToken(creds, doFetch, now)
        guard let token else { throw GeminiError(message: "oauth_creds.json 缺少 access_token 且无法刷新") }

        // loadCodeAssist 失败不阻塞配额查询
        let status: CodeAssistStatus
        do {
            status = try await loadCodeAssistStatus(token, doFetch)
        } catch {
            status = CodeAssistStatus(tier: nil, projectId: nil)
        }
        let projectId: String?
        if let known = status.projectId {
            projectId = known
        } else {
            projectId = await discoverProjectId(token, doFetch)
        }
        let buckets = try await retrieveUserQuota(token, projectId, doFetch)
        if buckets.isEmpty { throw GeminiError(message: "配额响应缺少 buckets 数据") }

        // 每个模型取最低 remaining_fraction 的桶（通常 input 侧先到限）
        var entries: [(modelId: String, fraction: Double, resetAt: Double?)] = []
        for raw in buckets {
            guard let bucket = raw.obj else { continue }
            guard let modelId = quotaStr(bucket["model_id"]),
                  let fraction = quotaNum(bucket["remaining_fraction"]) else { continue }
            if let idx = entries.firstIndex(where: { $0.modelId == modelId }) {
                if fraction < entries[idx].fraction {
                    entries[idx] = (modelId, fraction, toEpochMs(bucket["reset_time"]))
                }
            } else {
                entries.append((modelId, fraction, toEpochMs(bucket["reset_time"])))
            }
        }
        var windows = entries.map { entry -> QuotaWindow in
            let used = min(100, max(0, (1 - entry.fraction) * 100))
            return QuotaWindow(key: entry.modelId, label: entry.modelId,
                               usedPercent: used, percentage: used, nextResetAt: entry.resetAt)
        }
        // 百分比降序，同分保持出现顺序（对齐 TS Map 插入序 + 稳定排序），只留前 6 个模型
        windows = windows.enumerated().sorted { a, b in
            let pa = a.element.percentage ?? 0
            let pb = b.element.percentage ?? 0
            return pa != pb ? pa > pb : a.offset < b.offset
        }.map(\.element)
        windows = Array(windows.prefix(6))

        return finalizeAccount(
            kind: .gemini, label: geminiLabel,
            planName: status.tier.flatMap { geminiTierNames[$0] ?? $0 },
            windows: windows, anySuccess: true, now: now)
    } catch {
        return finalizeAccount(kind: .gemini, label: geminiLabel,
                               error: errText(error), httpStatus: readHttpStatus(error).map(Double.init), now: now)
    }
}

public func geminiAccount(home: String, env: [String: String], doFetch: FetchLike, now: Double) async -> QuotaAccount {
    guard let creds = readGeminiCreds(home, env) else { return missingAccount(.gemini, geminiLabel, now) }
    return await fetchGeminiAccount(creds, doFetch, now)
}
